File: project3.py
#################################################################
## PROJECT 3
#################################################################
## IMPORTS

import numpy as np
import time
import csv
import matplotlib.pyplot as plt
from scipy.constants import Boltzmann as k

#################################################################
## CLASSES

class System():
    """This class handles the low-temperature behavior of a 
    lattice material"""

    def __init__(self, N: int, J: float = -k, B: float = 0, T: float = 2) -> None:
        """
        parameters:
            N: lattice size [int]
            J: coupling strength [float] (usully on the order of kB)
            B: external magnetic field strngth [float]
            T: system temperature [float]
        """

        self.N = N
        self.J = J
        self.B = B
        self.T = T
        self.steps = 0
        self.accepted = 0

        # initializes the spin lattice with random spins
        self.lattice = np.random.choice([-1,1],size=(N,N))


    def magnetization(self) -> float:
        """calculates the magnetization of the lattice"""

        return np.sum(self.lattice)
    

    def hamiltonian(self) -> float:
        """calculates the energy of the lattice"""

        H = 0
        for i in range(self.N):
            for j in range(self.N):
                H += -self.J*self.lattice[i][j]*(self.lattice[(i+1)%self.N][j]+self.lattice[i-1][j]+self.lattice[i][(j+1)%self.N]+self.lattice[i][j-1])
                
                # this if-statement saves a lot of computing time when B=0
                if self.B:
                    H += -self.B*np.sum(self.lattice)
        return H


    def calc_dE(self, i: int, j: int) -> float:
        """calculates the change in energy from a spin flip at i,j"""

        return 2*self.J*self.lattice[i][j]*(self.lattice[(i+1)%self.N][j]+self.lattice[i-1][j]+self.lattice[i][(j+1)%self.N]+self.lattice[i][j-1])


    def step(self) -> None:
        """a step in the Metropolis algorithm using a random next site"""

        # random next site
        i, j = np.random.randint(self.N), np.random.randint(self.N)
        
        dE = self.calc_dE(i,j)

        # accept the move or not
        if dE < 0 or np.random.rand() < np.exp(-dE/k/self.T):
            self.accepted += 1
            self.lattice[i][j] = - self.lattice[i][j]
        
        self.steps += 1
            

    def equilibrialization(self, Nsweeps: int = 10) -> None:
        """sweeps for reaching thermal equilibrium
        the sweeps are not super intensive for N<=32"""

        for _ in range(Nsweeps):
            # do sweep
            for i in range(self.N):
                for j in range(self.N):
                    dE = self.calc_dE(i,j)
                    if dE < 0 or np.random.rand() < np.exp(-dE/k/self.T):
                        self.lattice[i][j] = - self.lattice[i][j]


    def show_state(self) -> None:
        """shows a figure of the state with binary spins"""
        
        plt.imshow(self.lattice,cmap="Greys")
        ax = plt.gca()
        # ax.set_xticks(np.arange(0, self.N, 1))
        # ax.set_yticks(np.arange(0, self.N, 1))

        ax.set_xticks([])
        ax.set_yticks([])

        # Labels for major ticks
        # ax.set_xticklabels(np.arange(1, self.N+1, 1))
        # ax.set_yticklabels(np.arange(1, self.N+1, 1))

        # Minor ticks
        ax.set_xticks(np.arange(-.5, self.N, 1), minor=True)
        ax.set_yticks(np.arange(-.5, self.N, 1), minor=True)

        # Gridlines based on minor ticks
        ax.grid(which='minor', color='grey', linestyle='dotted', linewidth=0.5)

        # Remove minor ticks
        ax.tick_params(which='minor', bottom=False, left=False)

        plt.xlabel("i")
        plt.ylabel("j")
        plt.title(f"Square Ising model with N={self.N}")
        plt.show()

#################################################################
## FUNCTIONS

def estimate_order(N,J,T,steps,number=10):

    order = []

    times = [time.time()]*2

    for j in range(number):
        system = System(N,J,T)
        for i in range(steps):

            times.pop(0)
            times.append(time.time())
            print(f"\r[{'#'*round((i+j*steps)/(steps*number -1)*20):.<20}] {round((i+j*steps)/(steps*number-1)*100):02}% |{round((times[1]-times[0])*(steps*number-1-i-j*steps)):04}s|",end="\r")
            system.step()

        order.append(abs(system.magnetization()))

    print("\nDone")

    return np.mean(order)    


def order_over_T(N,J,steps):

    Ts = [1,5,10,20,40,50,80,100]

    orders = [estimate_order(N,J,T,steps) for T in Ts]

    plt.plot(Ts,orders)
    plt.show()


def LLsizes(sizes: list[int] = [4,8,16,32], n_avgs: int = 100,
            flags: list[bool] = [True, False, False, False]) -> None:
    
    """Observe that this function is also written in C++"""

    Magnetization_flag = flags[0]
    Susceptibility_flag = flags[1]
    Specific_flag = flags[2]
    Cumulant_flag = flags[3]

    Ts = np.linspace(1,4,20)

    times = [time.time(), time.time()]
    
    magtot = [[] for _ in range(len(sizes))]
    sus_tot =  [[] for _ in range(len(sizes))]
    energy_tot = [[] for _ in range(len(sizes))]
    specific_tot = [[] for _ in range(len(sizes))]
    cumul_tot = [[] for _ in range(len(sizes))]

    for t,Ti in enumerate(Ts):

        # averages over iterations
        if Magnetization_flag or Susceptibility_flag:
            mags = [[] for _ in range(len(sizes))]
        if Susceptibility_flag:
            sus = [[] for _ in range(len(sizes))]
        if Specific_flag:
            energys = [[] for _ in range(len(sizes))]
            specifics = [[] for _ in range(len(sizes))]
        if Cumulant_flag:
            cumulants = [[] for _ in range(len(sizes))]

        for j in range(n_avgs):
            times.pop(0)
            times.append(time.time())
            print(f"\r[{'#'*round((j+t*n_avgs)/(n_avgs*len(Ts) -1)*20):.<20}] {round((j+t*n_avgs)/(n_avgs*len(Ts)-1)*100):03}% |{round((times[1]-times[0])*(n_avgs*len(Ts)-1-j-t*n_avgs)):04}s|",end="\r")

            # averages over steps
            if Magnetization_flag or Susceptibility_flag:
                temps_mag = [[] for _ in range(len(sizes))]
                if Susceptibility_flag:
                    temps_sq = [[] for _ in range(len(sizes))]
            if Specific_flag:
                temps_E = [[] for _ in range(len(sizes))]
                temp_spec = [[] for _ in range(len(sizes))]
            if Cumulant_flag:
                temps_m2 = [[] for _ in range(len(sizes))]
                temps_m4 = [[] for _ in range(len(sizes))]

            systems = [System(size,T=Ti) for size in sizes]

            # lets the system approach thermal equilibrium
            for system in systems:
                system.equilibrialization()

            steps = 40

            ## runs steps in the simulation to sample data
            ########
            for step in range(steps):
                for s in range(len(sizes)): # temp,system in zip(temps,systems):
                    systems[s].equilibrialization(Nsweeps=4)
                    if Magnetization_flag or Susceptibility_flag:
                        M = abs(systems[s].magnetization())
                        temps_mag[s].append(M)
                        if Susceptibility_flag:
                            temps_sq[s].append(M**2)
                    if Specific_flag:
                        E = systems[s].hamiltonian()
                        temps_E[s].append(E)
                        temp_spec[s].append(E**2)
                    if Cumulant_flag:
                        M = abs(systems[s].magnetization())
                        temps_m2[s].append(M**2)
                        temps_m4[s].append(M**4)
            ########

            ## averages the samples taken during the steps
            ########
            if Magnetization_flag or Susceptibility_flag:
                for s in range(len(sizes)):
                    mags[s].append(np.mean(temps_mag[s]))
                    if Susceptibility_flag:
                        sus[s].append((np.mean(temps_sq[s])-np.mean(temps_mag[s])**2)/systems[s].N**2)
            if Specific_flag:
                for s in range(len(sizes)):
                    energys[s].append(np.mean(temps_E[s])/k)
                    specifics[s].append(np.mean((temp_spec[s])-np.mean(temps_E[s])**2)/k)
            if Cumulant_flag:
                for s in range(len(sizes)):
                    cumulants[s].append(1-np.mean(temps_m4[s])/(3*np.mean(temps_m2[s])**2))
            ########

        ## averages over a number of realizations
        ########
        if Magnetization_flag or Susceptibility_flag:
            for s in range(len(sizes)):
                magtot[s].append(np.mean(mags[s]))
                if Susceptibility_flag:
                    sus_tot[s].append(np.mean(sus[s]))
        if Specific_flag:
            for s in range(len(sizes)):
                energy_tot[s].append(np.mean(energys[s]))
                specific_tot[s].append(np.mean(specifics[s]))
        if Cumulant_flag:
            for s in range(len(sizes)):
                cumul_tot[s].append(np.mean(cumulants[s]))


    print("\nDone")

    # print(system32.accepted/system32.steps)
    plotting_LLsize(magtot, sus_tot, specific_tot,cumul_tot, sizes, Ts)


def plotting_LLsize(Magnet, Suscept, Specific, Cumul, sizes: list[int], Ts:list[float]):
    def __ferro(T):
        return 1/(T-Tc)
    
    Tc = 2.2691853
    x = np.linspace(Tc+0.1,Ts[-1],200)
    ferros = __ferro(x)

    labels = [f"N={size}"for size in sizes]
    ms = ["^","s","o","D"]
    if len(Magnet[0])>0:
        plt.figure(1)
        plt.vlines([Tc],0,1,linestyles="--", colors="grey", label=r"$T_c$")
        for i,magt in enumerate(Magnet):
            plt.plot(Ts,magt,color = f"C{i}",linestyle="--",marker=ms[i],markerfacecolor="none",label=labels[i])
            # plt.scatter(Ts,magt,facecolors = "none",edgecolors = f"C{i}",marker=["o","s"][i%2],label=labels[i])
        plt.grid(linestyle="--")
        plt.legend()
        plt.title("Magnetization over temperature for different grid sizes")
        plt.ylabel(r"Magnetization $|M|/N^2$")
        plt.xlabel(r"Temperature $Tk_B/J$")
    if len(Suscept[0])>0:
        plt.figure(2)
        plt.vlines([Tc],0,max(Suscept[-1]),linestyles="--", colors="grey", label=r"$T_c$")
        for i,suscept in enumerate(Suscept):
            plt.plot(Ts,suscept,color = f"C{i}",linestyle="--",marker=ms[i],markerfacecolor="none",label=labels[i])
            # plt.scatter(Ts,suscept,facecolors = "none",edgecolors = f"C{i}",marker=["o","s"][i%2],label=labels[i])
        plt.grid(linestyle="--")
        plt.legend()
        plt.title("Susceptibility over temperature for different grid sizes")
        plt.ylabel(r"Susceptibility $\chi/N^2$")
        plt.xlabel(r"Temperature $Tk_B/J$")
        # plt.plot(x,ferros)
    if len(Specific[0])>0:
        plt.figure(3)
        plt.vlines([Tc],0,max(Specific[-1]),linestyles="--", colors="grey", label=r"$T_c$")
        for i,specific in enumerate(Specific):
            plt.plot(Ts,specific,color = f"C{i}",linestyle="--",marker=ms[i],markerfacecolor="none",label=labels[i])
            # plt.scatter(Ts,specific,facecolors = "none",edgecolors = f"C{i}",marker=["o","s"][i%2],label=labels[i])
        plt.grid(linestyle="--")
        plt.legend()
        plt.title("Specific heat over temperature for different grid sizes")
        plt.ylabel(r"Specific heat $C_B/k_BN^2$")
        plt.xlabel(r"Temperature $Tk_B/J$")
    if len(Cumul[0])>0:
        plt.figure(4)
        plt.vlines([Tc],min(Cumul[-1]),1.2*max(Cumul[-1]),linestyles="--", colors="grey", label=r"$T_c$")
        for i,cum in enumerate(Cumul):
            plt.plot(Ts,cum,color = f"C{i}",linestyle="--",marker=ms[i],markerfacecolor="none",label=labels[i])
            # plt.scatter(Ts,cum,facecolors = "none",edgecolors = f"C{i}",marker=["o","s"][i%2],label=labels[i])
        plt.grid(linestyle="--")
        plt.legend()
        plt.title("Cumulant over temperature for different grid sizes")
        plt.ylabel(r"Cumulant $U^4_L$")
        plt.xlabel(r"Temperature $Tk_B/J$")
    plt.show()


def read_csv(filename):
    T = []
    M = []
    X = []
    E = []
    Cv = []
    U4 = []

    with open(filename, "r") as csvfile:
        reader = csv.reader(csvfile)

        next(reader)

        for row in reader:
            T.append(float(row[0]))
            M.append(float(row[1]))
            X.append(float(row[2]))
            E.append(float(row[3]))
            Cv.append(float(row[4]))
            U4.append(float(row[5]))
    return T,M,X,E,Cv,U4

def plot_CPP(sizes):

    Ms = []
    Xs = []
    Es = []
    Cvs = []
    U4s = []

    for size in sizes:
        filename = f"Ferro_80T/output_N{size}.csv"
        T,M,X,E,Cv,U4 = read_csv(filename)
        Ms.append(M)
        Xs.append(X)
        Es.append(E)
        Cvs.append(Cv)
        U4s.append(U4)
    
    plotting_LLsize(Ms,Xs,Cvs,U4s,sizes,T)


def show_evolution():

    sys = System(32,-k,0,1)

    n_steps = 200000

    for i in range(n_steps):
        sys.step()
        if i in [0,0.5*n_steps,n_steps-1]:
    # sys.equilibrialization(Nsweeps=100)
    
            sys.show_state()

#################################################################
## MAIN

def main():

    # LLsizes(flags=[False,True,False,False], n_avgs=2)

    sizes = [4,8,16,32]

    plot_CPP(sizes)


#################################################################
## RUN CODE

if __name__ == "__main__":
    main()

File: project3.cpp
#include <random>
#include <cmath>
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

const double k = 1.3806503e-23;

class System {
public:
    // Member variables
    int N;
    double J;
    double B;
    double T;

    int **lattice;

    // Constructor
    System(); // Default constructor
    System(int N_in, double J_in, double B_in, double T_in);

    double magnetization();
    double hamiltonian();
    void sweep(int Nsweeps);
    void step();
    void setup(int N_in, double J_in, double B_in, double T_in);
    
private:
    double calc_dE(int i, int j);
};

System::System(){}

System::System(int N_in, double J_in, double B_in, double T_in) {

    N = N_in;
    J = J_in;
    B = B_in;
    T = T_in;

    lattice = new int*[N];
    for (int i = 0; i < N; ++i) {
        lattice[i] = new int[N];
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<double> dist(0, 1);

    // Initialize 2D lattice
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            double rand = dist(gen);
            int spin = (rand < 0.75) ? 1 : -1;

            lattice[i][j] = spin;
        }
    }

}

void System::setup(int N_in, double J_in, double B_in, double T_in)
{
    N = N_in;
    J = J_in;
    B = B_in;
    T = T_in;

    double frac;

    if (J_in > 0)
    {
        frac = 0.75;
    }
    else
    {
        frac = 0.5;
    }

    lattice = new int*[N];
    for (int i = 0; i < N; ++i) {
        lattice[i] = new int[N];
    }

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<double> dist(0, 1);

    // Initialize 2D lattice
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            // int spin = 1;
            double rand = dist(gen);
            int spin = (rand < frac) ? 1 : -1;

            // int spin = dist(gen);
            // if (spin == 0)
            // {
            //     spin = -1;
            // }

            lattice[i][j] = spin;
        }
    }
}

double System::magnetization()
{
    int total_M = 0;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            total_M = total_M + lattice[i][j];
        }
    }

    return total_M/(static_cast<double>(N)*static_cast<double>(N));
}

double System::hamiltonian()
{
    double H = 0.;
    

    for (int i = 0; i < N; ++i) 
    {
        for (int j = 0; j < N; ++j) 
        {
            int i_down = i;
            int j_down = j;
            
            if (i == 0)
            {
                i_down = N-1;
            }
            if (j == 0)
            {
                j_down = N-1;
            }

            H = H - J*lattice[i][j]*(lattice[(i+1)%N][j]+lattice[i_down][j]+lattice[i][(j+1)%N]+lattice[i][j_down]);
        }
    }

    if (B != 0)
    {
        double tot_spin = 0;

        for (int i = 0; i < N; ++i) 
        {
            for (int j = 0; j < N; ++j) 
            {
                tot_spin += 1.*lattice[i][j];
            }
        }
        H = H-B*tot_spin;
    }
    
    return H;
}

double System::calc_dE(int i, int j)
{

    int i_down = i;
    int j_down = j;

    if (i == 0)
    {
        i_down = N-1;
    }
    if (j == 0)
    {
        j_down = N-1;
    }

    return 2.*J*lattice[i][j]*(lattice[(i+1)%N][j]+lattice[i_down][j]+lattice[i][(j+1)%N]+lattice[i][j_down]) + 2.*B*lattice[i][j];
}

void System::step()
{
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dist(0, N-1);
    std::uniform_real_distribution<double> dist2(0, 1);

    int i = dist(gen);
    int j = dist(gen);

    double dE = calc_dE(i,j);

    if (dE < 0 || dist2(gen)<std::exp(-dE/k/T))
    {
        lattice[i][j] = -lattice[i][j];
    }
}

void System::sweep(int Nsweeps)
{

    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<double> dist2(0, 1);

    for (int i = 1; i <= Nsweeps; i++)
    {
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                
                double dE = calc_dE(i,j);

                if (dE < 0 || dist2(gen)<std::exp(-dE/k/T))
                    {
                        lattice[i][j] = -lattice[i][j];
                    }
            }
        }
    }
}

double* linspace(double low, double high, int steps)
{
    double* arr = new double[steps];

    int i = 0;
    for (double T = low; T<=high; T = T+(high-low)/static_cast<double>(steps))
    {
        arr[i] = T;
        i++;
    }

    return arr;
}

void loading(int j, int total)
{

    double frac = static_cast<double>(j)/static_cast<double>(total);

    int percentage = frac*100 + 0.5;

    std::cout << "[";

    for (int i = 0; i < frac*20; i++)
    {
        std::cout << "#";
    }
    for (int i = 0; i <= (1-frac)*20; i++)
    {
        std::cout << "-";
    }

    std::cout << "] [" << percentage << "%]\r";
    std::cout.flush();
}

void LLsizes(double J = k, double B = 0., int n_real = 10, int n_samples = 10, int T_steps = 100)
{
    int sizes[] = {4,8,16,32};
    int num_size = 4;

    double low = 1.;
    double high = 4.;

    double* Ts = linspace(low,high,T_steps);

    double mag_tot[num_size][T_steps];
    double sus_tot[num_size][T_steps];
    double energy_tot[num_size][T_steps];
    double specific_tot[num_size][T_steps];
    double cumul_tot[num_size][T_steps];

    for (int i = 0; i < T_steps; i++)
    {
        double mags[num_size];
        double sus[num_size];
        double energy[num_size];
        double specific[num_size];
        double cumul[num_size];

        for (int s = 0; s<num_size;s++)
        {
            mags[s] = 0.;
            sus[s] = 0.;
            energy[s] = 0.;
            specific[s] = 0.;
            cumul[s] = 0.;
        }

        for (int j = 0; j < n_real; j++)
        {

            loading(i*n_real+j,T_steps*n_real);

            double temp_mag[num_size];
            double temp_mag2[num_size];
            double temp_E[num_size];
            double temp_spec[num_size];
            double temp_mag4[num_size];

            // System systems[] = {System(4,k,0,Ts[i]),System(8,k,0,Ts[i]),System(16,k,0,Ts[i]),System(32,k,0,Ts[i])};

            System systems[num_size];

            for (int s = 0; s<num_size;s++)
            {
                systems[s].setup(sizes[s],J,B,Ts[i]);
                systems[s].sweep(200);
                temp_mag[s] = 0.;
                temp_mag2[s] = 0.;
                temp_mag4[s] = 0.;
                temp_E[s] = 0.;
                temp_spec[s] = 0.;
            }

            for (int l = 0; l < n_samples; l++)
            {
                for (int s = 0; s<num_size;s++)
                {
                    double M = std::abs(systems[s].magnetization());
                    double E = std::abs(systems[s].hamiltonian());
                    // systems[s].sweep(2);
                    double E2 = std::pow(systems[s].hamiltonian(),2);
                    double M2 = std::abs(std::pow(systems[s].magnetization(),2));
                    // systems[s].sweep(2);
                    double M4 = std::abs(std::pow(systems[s].magnetization(),4));

                    temp_mag[s] += M;
                    temp_mag2[s] += M2;
                    temp_E[s] += E;
                    temp_spec[s] += E2;
                    temp_mag4[s] += M4;

                    systems[s].sweep(2);
                }
            }

            // averages over samples
            for (int s = 0; s<num_size;s++)
            {
                double M_mean = temp_mag[s]/static_cast<double>(n_samples);
                double M2_mean = temp_mag2[s]/static_cast<double>(n_samples);
                double M4_mean = temp_mag4[s]/static_cast<double>(n_samples);
                double E_mean = temp_E[s]/static_cast<double>(n_samples);
                double E2_mean = temp_spec[s]/static_cast<double>(n_samples);

                mags[s] += M_mean;
                sus[s] += (M2_mean-std::pow(M_mean,2))/(Ts[i]);
                energy[s] += E_mean/k;
                specific[s] += (E2_mean-std::pow(E_mean,2))/(k*k);
                cumul[s] += 1.-M4_mean/(3.*std::pow(M2_mean,2));
            }

        }

        // averages over realizations
        for (int s = 0; s<num_size;s++)
        {
            mag_tot[s][i] = mags[s]/static_cast<double>(n_real);
            sus_tot[s][i] = sus[s]/static_cast<double>(n_real);
            energy_tot[s][i] = energy[s]/static_cast<double>(n_real);
            specific_tot[s][i] = specific[s]/static_cast<double>(n_real);
            cumul_tot[s][i] = cumul[s]/static_cast<double>(n_real);
        }
    }

    for (int s = 0; s < num_size; s++)
    {
        std::string filename = "output_N" + std::to_string(sizes[s]) + ".csv";

        std::ofstream file(filename);

        file << "T,M,X,E,Cv,U4\n";

        for (int i=0; i<T_steps;i++)
        {   
            file << Ts[i];
            file << ","; 
            file << mag_tot[s][i];
            file << ","; 
            file << sus_tot[s][i];
            file << ","; 
            file << energy_tot[s][i];
            file << ","; 
            file << specific_tot[s][i];
            file << ","; 
            file << cumul_tot[s][i];
            file << "\n";
        }
        

        file.close();
    }    

}

int main()
{
    
    int n_real = 20;
    int n_samples = 1000;
    int T_steps = 40;

    double B = 0;
    double J = -k;

    LLsizes(J,B,n_real,n_samples,T_steps);
    
    return 0;
}